molecule/provisioners/libvirtprovisioner.py

```diff
@@ -0,0 +1,923 @@
+        except libvirt.libvirtError:
+            return self._define_pool(pool_name)
+
+    def _define_network(self, network=None):
+        """
+        Define a libvirt network and start it.
+        TODO: Make this all configurable. 'network' will be a dict that describes a libvirt network, populated based on molecule.yml.
+        libvirt:
+            networks:
+                - name: molecule0
+                  forward: nat|none
+                  #bridge: virbr10
+                  cidr: 192.168.121.1/24
+        """
+        cidr_net = netaddr.IPNetwork(network['cidr'])
+        utilities.print_info("Creating libvirt network: {}".format(network[
+            'name']))
+        net = ET.Element('network', ipv6='no')
+        ET.SubElement(net, 'name').text = network['name']
+        if 'forward' in network and network['forward'] != 'none':
+            forward = ET.SubElement(net, 'forward', mode='nat')
+            nat = ET.SubElement(forward, 'nat')
+            port = ET.SubElement(nat, 'port', start='1024', end='65535')
+        if 'bridge' in network:
+            bridge = ET.SubElement(
+                net, 'bridge', name=network['bridge'], stp='on', delay='0')
+        ip = ET.SubElement(
+            net, 'ip', address=str(cidr_net.ip), netmask=str(cidr_net.netmask))
+        want_dhcp = network.get('dhcp', True)
+        if want_dhcp:
+            dhcp = ET.SubElement(ip, 'dhcp')
+            dhcprange = ET.SubElement(
+                dhcp,
+                'range',
+                start=str(cidr_net.ip),
+                end=str(list(cidr_net)[-2]))
+
+        netxml = ET.tostring(net)
+        LOG.debug("\tXMLDesc: {}".format(netxml))
+        newnet = self._libvirt.networkDefineXML(netxml)
+        return newnet.create()
+
+    def _up_network(self, network=None):
+        """
+        Up or define/up a libvirt network.
+
+        :param: a dict describing the network, with unique 'name' and 'cidr' keys
+        """
+        # TODO: Remove this when we have molecule defaults for libvirt networks
+        if not network:
+            network = {'name': 'molecule0',
+                       'forward': 'nat',
+                       # 'bridge': 'virbr10',
+                       'cidr': '192.168.122.1/24'}
+
+        try:
+            net_found = self._libvirt.networkLookupByName(network['name'])
+            if net_found and not net_found.isActive():
+                net_found.create()
+            return net_found  # existing network is now running
+        except libvirt.libvirtError:
+            return self._define_network(network)
+
+    def _build_domain_xml(self, instance):
+        """
+        Builds a string of xml suitable for self._libvirt.defineXML(xml)
+
+        :param: a dict that describes an instance (a virtualization guest that this provisioner should create)
+        :return: an xml string
+        """
+
+        # Basic elements
+        dom = ET.Element('domain', type='kvm')
```

Comment:
Is all this really the best way to build this XML?  How about having a couple of templates and interpolating/combining them?  This way of building XML is okay-ish but it's not really clear what the output should be.  Using templates would be clearer.